src/ncp/ncp_host.hpp

```diff
@@ -63,7 +63,9 @@ class NcpHost : public MainloopProcessor, public ThreadHost
     ~NcpHost(void) override = default;
 
     // ThreadHost methods
-    void            GetDeviceRole(const DeviceRoleHandler aHandler) override;
+    void GetDeviceRole(const DeviceRoleHandler aHandler) override;
+    void Join(const otOperationalDatasetTlvs &aActiveOpDatasetTlvs, const AsyncResultReceiver aReceiver) override;
+    void Leave(const AsyncResultReceiver aReceiver) override;
```

Comment:
For the receiver parameters, I'd suggest to either pass by const reference or pass by value.